OpenSim/Common/Delay.h

```diff
@@ -0,0 +1,179 @@
+ *
+ * @see ScalarDelay, VectorDelay
+ *
+ *
+ * This Component has a single input named "input" and a single output named
+ * "output". The typical way to use this Component is as a subcomponent in
+ * another component. Here is a basic example in which a controller specifies
+ * the control signal for a single actuator using the delayed value of a
+ * coordinate:
+ * @code
+ * using namespace OpenSim;
+ * using namespace SimTK;
+ * class MyController : Controller {
+ * OpenSim_DECLARE_CONCRETE_OBJECT(MyController, Controller);
+ * public:
+ *     OpenSim_DECLARE_PROPERTY(delay, double, "Duration of delay (seconds).");
+ *     MyController() { constructProperties(); }
+ *     void computeControls(const State& s, Vector &controls) const override {
+ *         double q = _coordDelay.getValue(s);
+ *         // Delayed negative feedback control law.
+ *         double u = -10 * q;
+ *         getModel().getActuatorSet()[0].addInControls(Vector(1, u), controls);
+ *     }
+ * private:
+ *     void constructProperties() {
+ *          constructProperty_delay(0.01); // 10 milliseconds
+ *     }
+ *     void extendFinalizeFromProperties() override {
+ *         _coordDelay.set_delay(get_delay());
+ *         addComponent(&_coordDelay);
+ *     }
+ *     void extendConnectToModel(Model& model) override {
+ *         Super::extendConnectToModel(model);
+ *         const auto& coord = model.getCoordinateSet()[0];
+ *         // Set the input for the Delay component.
+ *         _coordDelay.getInput("input").connect(coord.getOutput("value"));
+ *     }
+ *
+ *     ScalarDelay _coordDelay;
+ * };
+ * @endcode
+ *
+ * This class is implemented via a SimTK::Measure_<T>::Delay.
+ *
+ * @ingroup operators
+ *
+ */
+template<typename T>
+class Delay : public Component {
+OpenSim_DECLARE_CONCRETE_OBJECT_T(Delay, T, Component);
+public:
+
+    /** @name Property declarations
+    These are the serializable properties associated with a Delay. **/
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(delay, double,
+        "The duration (nonnegative, seconds) by which the output is delayed.");
+    /**@}**/
+
+    /// Default constructor.
+    Delay();
+
+    /// Convenience constructor that sets the delay property.
+    explicit Delay(double delay);
+
+    /// Get the delayed value (the input's value at time t-delay).
+    T getValue(const SimTK::State& s) const;
+
+private:
+
+    void constructProperties() override;
+    void constructInputs() override;
+    void constructOutputs() override;
+
+    void extendFinalizeFromProperties() override;
+    void extendAddToSystem(SimTK::MultibodySystem& system) const override;
+
+    SimTK::MeasureIndex _delayMeasureIndex;
+};
+
+template<class T>
+Delay<T>::Delay() {
+    constructInfrastructure();
+}
+
+template<class T>
+Delay<T>::Delay(double delay) : Delay() {
+    set_delay(delay);
+}
+
+template<class T>
+void Delay<T>::constructProperties() {
+    constructProperty_delay(0.0);
+}
```

Comment:
> how does the DelayMeasure treat a 0 delay

I don't see any special handling of delay==0 in the Measure::Delay implementation. By default the measure is prohibited from looking at the current value of the source (delayed) measure, because we assume that the current source value depends on the delayed value. So it would have to extrapolate from previous values to get an estimate of the current value, rather than just evaluating it. There is a method `setCanUseCurrentValue()` that overrides that default; in that case I think this would be a very slow way to pass the value through.

I don't think Delay has every been tried with delay==0 so it is very possible it would fail. Please file an issue if it does!